### PR TITLE
Avoid lost CompilationEvent instances on analysis cancellation

### LINKED ISSUE #46950: Flaky test - ErrorSquiggleProducerTests.ErrorDoesNotCrashPastEOF
Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Squiggles.ErrorSquiggleProducerTests.ErrorDoesNotCrashPastEOF

PR: #46946
Build: https://dev.azure.com/dnceng/public/_build/results?buildId=777134&view=ms.vss-test-web.build-test-results-tab&runId=24314366&resultId=120403&paneView=debug

Stack trace:
```
Assert.Equal() Failure\r\nExpected: 3\r\nActual:   0
 en Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) en C:\projects\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:línea 40
   en Xunit.Assert.Equal[T](T expected, T actual) en C:\projects\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:línea 24
   en Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Squiggles.ErrorSquiggleProducerTests.<ErrorDoesNotCrashPastEOF>d__4.MoveNext() en /_/src/EditorFeatures/CSharpTest/Squiggles/ErrorSquiggleProducerTests.cs:línea 201
--- Fin del seguimiento de la pila de la ubicación anterior donde se produjo la excepción ---
   en System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   en System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   en Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() en C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:línea 264
--- Fin del seguimiento de la pila de la ubicación anterior donde se produjo la excepción ---
   en System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   en System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   en Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext() en C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:línea 48
--- Fin del seguimiento de la pila de la ubicación anterior donde se produjo la excepción ---
   en System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   en System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   en Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() en C:\projects\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:línea 90
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #46950

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs [hunk 1]
```diff
@@ -107,15 +107,21 @@ private PerAnalyzerState GetAnalyzerState(DiagnosticAnalyzer analyzer)
...
+            AsyncQueue<CompilationEvent> eventQueue,
+            ImmutableArray<AdditionalText> additionalFiles,
+            AnalyzerDriver driver,
+            CancellationToken cancellationToken)
         {
             try
             {
                 await EnsureAnalyzerActionCountsInitializedAsync(driver, cancellationToken).ConfigureAwait(false);
 
                 using (_gate.DisposableWait(cancellationToken))
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs:121]: 📝 When cancellation occurred here, the `compilationEvents` collection was getting discarded.

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs:121]: Nice!

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews

**Comment by jcouv** [comment]
Is there a way to add a test for original scenario from issue, or is the timing too tricky?

**Comment by sharwell** [comment]
@jcouv I think it would be very difficult to reproduce this in a test

**Pull request merged**